sdk/identity/Azure.Identity/src/TokenCache.cs

```diff
@@ -0,0 +1,289 @@
+            }
+            else
+            {
+                stream.Read(data, 0, data.Length);
+            }
+
+            return new TokenCache(data);
+        }
+
+        internal virtual async Task RegisterCache(bool async, ITokenCache tokenCache, CancellationToken cancellationToken)
+        {
+            if (async)
+            {
+                await _lock.WaitAsync(cancellationToken).ConfigureAwait(false);
+            }
+            else
+            {
+                _lock.Wait(cancellationToken);
+            }
+
+            try
+            {
+                if (!_cacheAccessMap.TryGetValue(tokenCache, out _))
+                {
+                    tokenCache.SetBeforeAccessAsync(OnBeforeCacheAccessAsync);
+
+                    tokenCache.SetAfterAccessAsync(OnAfterCacheAccessAsync);
+
+                    _cacheAccessMap.Add(tokenCache, new CacheTimestamp());
+                }
+            }
+            finally
+            {
+                _lock.Release();
+            }
+        }
+
+        private async Task OnBeforeCacheAccessAsync(TokenCacheNotificationArgs args)
+        {
+            if (_disposedValue)
+            {
+                throw new ObjectDisposedException(nameof(TokenCache));
+            }
+
+            await _lock.WaitAsync().ConfigureAwait(false);
+
+            try
+            {
+                args.TokenCache.DeserializeMsalV3(_data, true);
+
+                _cacheAccessMap.GetOrCreateValue(args.TokenCache).Update();
+            }
+            finally
+            {
+                _lock.Release();
+            }
+        }
+
+
+        private async Task OnAfterCacheAccessAsync(TokenCacheNotificationArgs args)
+        {
+            if (_disposedValue)
+            {
+                throw new ObjectDisposedException(nameof(TokenCache));
+            }
+
+            if (args.HasStateChanged)
+            {
+                await UpdateCacheDataAsync(args.TokenCache).ConfigureAwait(false);
+            }
+        }
+
+        private async Task UpdateCacheDataAsync(ITokenCacheSerializer tokenCache)
+        {
+            await _lock.WaitAsync().ConfigureAwait(false);
+
+            try
+            {
+                if (!_cacheAccessMap.TryGetValue(tokenCache, out CacheTimestamp lastRead) || lastRead.Value < _lastUpdated)
+                {
+                    _data = await MergeCacheData(_data, tokenCache.SerializeMsalV3()).ConfigureAwait(false);
+                }
+                else
+                {
+                    _data = tokenCache.SerializeMsalV3();
+                }
+
+                _cacheAccessMap.GetOrCreateValue(tokenCache).Update();
+
+                _lastUpdated = DateTime.UtcNow;
+            }
+            finally
+            {
+                _lock.Release();
+            }
+
+            if (Updated != null)
+            {
+                await Updated.Invoke(new TokenCacheUpdatedArgs(this)).ConfigureAwait(false);
```

Comment:
What if there are multiple registrations? You are not awaiting all of them.

Reply:
fixed